datalad/cmd.py

```diff
@@ -6,66 +6,152 @@
+    This can be enabled or disabled for both of them independently.
+    Additionally allows for dry runs. This is achieved by initializing the `Runner` with `dry=True`.
+    The Runner will then collect all calls as strings in `commands`.
     """
 
     __slots__ = ['commands', 'dry']
 
     def __init__(self, dry=False):
+        self.dry = dry
         self.commands = []
-        self.dry = dry                    # TODO: make use of it
 
-    # TODO -- remove dry_run specification here -- use constructor parameter
-    #def __call__(self, cmd, dry_run=False):
-    def getstatusoutput(self, cmd, dry_run=None):
-        """A wrapper around commands.getstatusoutput
+    def __call__(self, cmd, *args, **kwargs):
+        """Convenience method
+
+        This will call run() or call() depending on the kind of `cmd`.
+        If `cmd` is a string it is interpreted as the to be executed command.
+        Otherwise it is expected to be a callable.
+        Any other argument is passed to the respective method.
+
+        Parameters
+        ----------
+        cmd: str or callable
+           command string to be executed via shell or callable to be called.
+
+        *args and **kwargs:
+           see Runner.run() and Runner.call() for available arguments.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Raises
+        ------
+        ValueError if cmd is neither a string nor a callable.
         """
+
+        if isinstance(cmd, basestring):
+            self.run(cmd, *args, **kwargs)
+        elif callable(cmd):
+            self.call(cmd, *args, **kwargs)
+        else:
+            raise ValueError("Argument 'command' is neither a string nor a callable.")
+
+    def run(self, cmd, log_stdout=False, log_stderr=True):
+        """Runs the command `cmd` using shell.
+
+        In case of dry-mode `cmd` is just added to `commands` and it is executed otherwise.
+        Allows for seperatly logging stdout and stderr  or streaming it to system's stdout
+        or stderr respectively.
+
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        log_stdout: bool
+            If True, stdout is logged. Goes to sys.stdout otherwise.
+
+        log_stderr: bool
+            If True, stderr is logged. Goes to sys.stderr otherwise.
+
+        Returns
+        -------
+        Status code as returned by the called command or `None` in case of dry-mode.
+
+        Raises
+        ------
+        RunTimeError
+           if command's exitcode wasn't 0 or None
+        """
+
+        if log_stdout:
+            outputstream = subprocess.PIPE
+        else:
+            outputstream = sys.stdout
+
+        if log_stderr:
+            errstream = subprocess.PIPE
+        else:
+            errstream = sys.stderr
+
```

Comment:
not for a change here but just a bit of pythonism:
```
outputstream = subprocess.PIPE if log_stdout else sys.stdout
```
saves code real estate and keystrokes

but to say the truth I am not yet using this construcct myself all the time